### DISCUSSION TITLE
How to: Export "intelligent" playlists for AzuraCast from your collection using Quod Libet

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6505] [Show and tell]
## Ever had the problem of regenerating many playlists for AzuraCast manually—from an ever-changing local collection of music?

Here’s how I currently do it.

### Use Quod Libet for searching

[Quod Libet](https://quodlibet.readthedocs.io/) has fantastic ideas behind it, but it is sluggish like hell on a 157,000 song collection (single-threaded, no database). Also, it insists in writing ID2v2.4 tags to MP3 (and other) files. So if you need ID3v2.3 tags, better _don’t use it for tagging_ (I exclusively use [MusicBrainz Picard](https://picard.musicbrainz.org/) for that purpose). You _can_ use Quod Libet’s advanced tagging features if your workflow allows ID3v2.4 in all places, or you only use FLAC.

Still, Quod Libet is the best tool for extremely complicated searches like
```
&(|(grouping=electronic,genre=|(electronic,berlin school),artist=|(klaus schulze,peter mergener,tomislav ocvirek)), !grouping=|(schräg,hip,rap,metal,hardcore,emo,classical,world,drum and bass,industrial,ebm,goth,dark,downtempo,funk,techno,rock,house,trance,new age,new wave,indie,humour,comedy,jazz,pop,disco,eurodance,tango,easy listening), !genre=|(schräg,hip,rap,metal,hardcore,emo,classical,world,drum and bass,idm,industrial,ebm,goth,dark,downtempo,funk,techno,dub,synth,hardstyle,rock,house,trance,new age,new wave,indie,humour,comedy,jazz,pop,disco,big beat,breakbeat,r&b,eurodance,tango,easy listening,lounge,bossa nova,latin,fusion,spoken,avant-garde,ndw,piano,club,a cappella), !artist=|(david guetta,frida gold,inna,lancer,madonna,arne pahlke,atmozfears,barbara morgenstern,wild eyes,björk,bob sinclar,brian eno,stu phillips), #(length>=2:00))
```
This is my actual search to arrive at an electronic music playlist I’d like to use with AzuraCast.

Quod Libet is also a great tool to [find problems in your collection](https://github.com/quodlibet/quodlibet/discussions/4349), like finding missing ReplayGain tags
```
&(~format=!wav,replaygain_track_gain="")
```
or simply checking which tracks have been modified (=are new) within the last week
```
#(mtime < 1 week)
```

Quod Libet can _save_ such searches, and you can reuse these further on, making "dynamic" or "intelligent" playlists a snap: You only have to update the library. No manual playlist creation anymore.

### Use `ql-playlists` to auto-generate all playlists

I wrote a little bash script called [`ql-playlists`](https://gist.github.com/Moonbase59/a62363d61a4dcb596bd27d464368a9a4) that you can use to make all Quod Libet saved searches into M3U playlists, ready to import into AzuraCast.

Thanks to AzuraCast’s fantastic _import playlist file matching_, you can even have paths in the M3U that aren’t _exactly_ the same as they’d look seen from the start of your station’s media folder. Thus, you can, for example, use relative paths in your playlist file and tuck it safely away, say in `~/Music/Playlists/Radio`, like I do.

Then just let `ql-playlists` fetch your Quod Libet searches and make them into playlists ready for uploading. This will take some time, so grab your favorite beverage and watch…

![matthias@e6510: ~-Musik-Playlists-Radio_001](https://github.com/AzuraCast/AzuraCast/assets/3706922/b5283def-53b1-45ee-889c-413313559adb)

Read more about `ql-playlists`, its features and usage in [my Gist](https://gist.github.com/Moonbase59/a62363d61a4dcb596bd27d464368a9a4?permalink_comment_id=4660429#gistcomment-4660429).

### Station media vs. large local collection

You must upload the media to your station _before_ you upload the playlist. You _should_ use the same folder structure locally and in your station’s `media` folder. In any case, you _must_ use the same filenames.

If your collection holds far more tracks than your station media folder (which is often the case), you could mirror the station media in, say `~/Music/Radio`, while your collection also has other folders like `~/Music/Tagged`, and all are in the Quod Libet library. Then simply use a search in Quod Libet that only includes the station media from the `~/Music/Radio` folder, such as
```
&(~filename=/\/Music\/Radio\//, ambient)
```
This would search only "ambient" music files having `/Music/Radio/` in their path. If you fear you might have an artist called `Music` with an album `Radio`, you can as well check for "starting with" an absolute path, like `/home/matthias/Music/Radio/`:
```
&(~filename=/^\/home\/matthias\/Music\/Radio\//, ambient)
```

You can even include _already saves searches_ in a search:
```
&(~filename=/^\/home\/matthias\/Music\/Radio\//, @(saved: Electronic))
```

Save it, for example, as "Radio - Electronic", and it will create a "Radio - Electronic" playlist with the search criteria I mentioned above, _but restricted to whatever media is already in the `/home/matthias/Music/Radio/` folder!_

**Hint:** Since playlists generated by `ql-playlists` are essentially just text files with a bunch of file paths in them, you could even use these to automate SFTP media file uploads to your station.

### Real-life workflow example

To illustrate all this, here are some screenshots from real life. I had just updated my music collection with several new albums, and now go and update my station playlists.

For any easy check if station playlists need updating, I just compare the number of songs. (It _could_ happen that you generate a new playlist with totally different songs that has the same number of entries, but that is rather unlikely.)

![Arbeitsfläche 1_019](https://github.com/AzuraCast/AzuraCast/assets/3706922/1c60b525-c5be-411e-beb0-2dcd5b11bb0d)

![Arbeitsfläche 1_020](https://github.com/AzuraCast/AzuraCast/assets/3706922/3b229a48-aa53-4c81-b724-6ee5efe2bbc7)

![Arbeitsfläche 1_021](https://github.com/AzuraCast/AzuraCast/assets/3706922/d3252e1a-1427-47d5-aa35-7ff4baa298cd)

---

**Note:** `ql-playlists` is for Linux. Might work on MacOS and Windows WSL, no guarantees. (Tested on current versions of Arch, Manjaro, Ubuntu, Linux Mint).

**Enjoy** and let me know if and how you use `ql-playlists`, or any errors you find.

If you find this useful, _upvote_ by using the blueish up-arrow button below this post, _not_ by using the :+1: emoji. (You can still do that, but it won’t upvote the discussion, just give me a warm feeling…)

**Comment by BusterNeece** [comment]
@Moonbase59 Should we be writing ID3v2.4 tags as well? Is that of benefit in this scenario?

**Reply by Moonbase59** [reply]
Hm, I might have been unclear. The benefit in this scenario is the fantastic ways of **searching** and **automating generation of always up-to-date playlists** the _Quod Libet_/_`ql-playlists`_ combo offers.

Quod Libet is unsurpassed when it comes to doing complicated searches. Despite its deficiencies, the possibilities it offers tapping one’s collection and automating things via the CLI (or Python) are just endless. Plus, they’re one of the few that handle seldom-used tags like _mood_ or _language_ (the song is sung in). If you need that, you need it badly…

Since above is based on saved _searches_, not static playlists, generated playlists for upload into the station _will always be up-to-date_—automatically. A main benefit if wanting to keep your station updated with new albums and songs. (And the only reason I wrote `ql-playlists`.)

I don’t currently let Quod Libet _write_ any single tag, because it would destroy my ID3v2.3.0-based collection by overwriting with ID3v2.4, and probably even _delete_ some tags I’d like to keep. (It might still be great in a sole ID3v2.4 + Vorbis Comments workflow.)

---

**Re tags needing to be written in AzuraCast:**

That should depend on the user’s setting. ID3v2.4 is more modern, understands UTF-8, and supports multiple (same) tags. Under Linux (fortunately?) ID3v2.4 has—more or less—become a de-facto standard, which may be good.

**But** still today ID3v2.4 is not understood by many players. My entire ~157k collection is _still_ tagged using ID3v2.3.0 (UTF-16) for compatibility reasons (as far as MP3 and other file formats using ID3 tags are concerned).

There _is_ a reason why I mount my "sacred" music collection (collected over almost 4 decades) only **read-only** further up in any audio chains. You wouldn’t believe the amount of corrupt ID2 and ID3 tags (even 3 or 4 sorts being _mixed_ in one MP3!) I’ve seen over the years, creating the most awkward problems in later processing. Think alone of ReplayGain: The uppercase vs. lowercase `TXXX` `replaygain` tags, iTunes variant, `RVAD` and `RVA2` and which to use, what priority if more than one ReplayGain system is used…—even today an unsolved problem! And the main reason why I wrote `loudgain` back in the days. (Fortunately, Liquidsoap reads those variants okay.)

**Here be dragons.** ID3 tagging is a mess. Really. Fortunately, with _Vorbis Comments_ (FLAC, OGG) all this is much easier, and almost any application does _that_ right.

Today, I’d always recommend an application software to…
- accept music files/collections to be read-only, without any adverse effects,
- _not_ write any tags without the user’s consent, if anyhow possible,
- let the user select between _ID3v2.3.0 with UTF-16 encoding_ and _ID3v2.4 with UTF-8 encoding_,
- only write UTF-16/UTF-8 tags to tags where it’s needed (like `TXXX` frames, for example),
- write tags like the `replaygain*` _always_ in ISO-8859-1 encoding.

I also usually **warn** people of modifying tags with arbitrary programs, or letting software do that without their consent. In my experience (and workflow) **_MusicBrainz Picard_ should always be the last step in tagging files** (because they do _everything_ right!), and then the file _never be touched again_, if anyhow possible.

**Reply by Moonbase59** [reply]
After writing this long answer, I realize you wrote "Is that _of_ benefit…" and not "Is that _the_ benefit…". Must be too early in the day for me… :rofl: 

No, this would work, with any file type and tag format Quod Libet and AzuraCast support. It’s just about being able to do elaborated searches and keeping station playlists up-to-date.

**Reply by BusterNeece** [reply]
This does have me looking into ways to support crazy searches. I think if you're willing to give up some initial performance on the cache-warming first run, we could definitely implement a query language of similar complexity.

**Reply by Moonbase59** [reply]
Yeah, well… Then again, it makes me think if we should have _another_ "trying-to-do-it-all" tool. And it would require _lots_ of odd tag handling (look at the tag mapping in Picard, as an example), multiple tag field (genre!) support with various separators (for ID3, again!)., etc.

Thinking about how badly most people do their tagging, plus the usually limited space on el-cheapo cloud VMs, we might assume most web radio stations only have (a very limited) part of their collection online. Not everyone is as crazy as me and keeps an NFS mount to their main music collection inside AzuraCast (and would use the cloud only for setting up relays).

I somehow think most web radio stations will probably upload only a rather small basic and well-curated rotation into their stations, thus probably eliminating the need for such overhead. Apart from that, in many countries you’ll have to pay a hefty surcharge on license fees if you have a large collection available in the radio. "Lying" to the authorities isn’t an option if you want to make song requests available.

It is tempting, I agree. And I _do_ sometimes miss advanced search functionality in AzuraCast, be it just searching for a song or actually to set up a playlist. But in the meantime I can live with what software like _Quod Libet_, _Clementine_/_Strawberry_ and _Guayadeque_ give me, and upload a playlist.

And I wonder if users might want to learn _yet another_ search syntax. It’s complicated enough already. Look at two real examples for my "Electronic" playlist.

The same search in _Clementine_/_Strawberry_ and _Quod Libet_:

```
(grouping:electronic OR genre:electronic OR genre:"berlin school" OR artist:"klaus schulze" OR artist:"peter mergener" OR artist:"tomislav ocvirek") -grouping:schräg -genre:schräg -grouping:hip -genre:hip -grouping:rap -genre:rap -grouping:metal -genre:metal -grouping:hardcore -genre:hardcore -grouping:emo -genre:emo -grouping:classical -genre:classical -grouping:world -genre:world -grouping:"drum and bass" -genre:"drum and bass" -genre:idm -grouping:industrial -genre:industrial -grouping:ebm -genre:ebm -grouping:goth -genre:goth -grouping:dark -genre:dark -grouping:downtempo -genre:downtempo -grouping:funk -genre:funk -grouping:techno -genre:techno -genre:dub -genre:synth -genre:hardstyle -grouping:rock -genre:rock -grouping:house -genre:house -grouping:trance -genre:trance -grouping:"new age" -genre:"new age" -grouping:"new wave" -genre:"new wave" -grouping:indie -genre:indie -grouping:humour -genre:humour -grouping:comedy -genre:comedy -grouping:jazz -genre:jazz -grouping:pop -genre:pop -grouping:disco -genre:disco -genre:"big beat" -genre:breakbeat -genre:"r&b" -grouping:eurodance -genre:eurodance -grouping:tango -genre:tango -grouping:"easy listening" -genre:"easy listening" -genre:lounge -genre:"bossa nova" -genre:latin -genre:fusion -genre:spoken -genre:"avant-garde" -genre:NDW -genre:piano -genre:club -genre:"a cappella" -artist:"david guetta" -artist:"frida gold" -artist:"inna" -artist:lancer -artist:madonna -artist:"arne pahlke" -artist:atmozfears -artist:"barbara morgenstern" -artist:"wild eyes" -artist:björk -artist:"bob sinclar" -artist:"brian eno" -artist:"stu phillips" length:>=2:00
```

```
&(|(grouping=electronic,genre=|(electronic,berlin school),artist=|(klaus schulze,peter mergener,tomislav ocvirek)), !grouping=|(schräg,hip,rap,metal,hardcore,emo,classical,world,drum and bass,industrial,ebm,goth,dark,downtempo,funk,techno,rock,house,trance,new age,new wave,indie,humour,comedy,jazz,pop,disco,eurodance,tango,easy listening), !genre=|(schräg,hip,rap,metal,hardcore,emo,classical,world,drum and bass,idm,industrial,ebm,goth,dark,downtempo,funk,techno,dub,synth,hardstyle,rock,house,trance,new age,new wave,indie,humour,comedy,jazz,pop,disco,big beat,breakbeat,r&b,eurodance,tango,easy listening,lounge,bossa nova,latin,fusion,spoken,avant-garde,ndw,piano,club,a cappella), !artist=|(david guetta,frida gold,inna,lancer,madonna,arne pahlke,atmozfears,barbara morgenstern,wild eyes,björk,bob sinclar,brian eno,stu phillips), #(length>=2:00))
```

(Large collections collected over decades _do_ have their downsides and pitfalls… :wink: )

**Would we really want monsters like that in AzuraCast? Or just concentrate on what it does best: Reliable _playout_?**

In any case, _should_ you implement such a thing, _please_ at least follow [Picard’s "unified" tag names and tag mapping](https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html), so people don’t have to remember yet another name for the same thing. For instance, `releasetype` should be called just that even if it is `RELEASETYPE` in Vorbis Comments, `TXXX:MusicBrainz Album Type` in ID3v2.3, `MUSICBRAINZ_ALBUMTYPE` in APEv2, `----:com.apple.iTunes:MusicBrainz Album Type` in iTunes MP4, and `MusicBrainz/Album Type` in ASF/WMA Windows Media. (Used to reliably find live albums/songs, for instance.)

Plus, of course, "free format" tag names must be possible, like in Vorbis Comments or ID3v2 `TXXX` frames.

And, ideally, [tags from Picard’s _last.fm.ng_ plugin](https://github.com/quodlibet/quodlibet/issues/4345), to tap the "swarm intelligence" for Decade ( `1970s`), Category ( `Cover`, `Female Vocalist`, …), City and Country of the artist’s origin, and Occasion (`Wedding`, `Funeral`, `Christmas`, …).

**Pandora’s Box**, remember? :wink:

**Reply by BusterNeece** [reply]
@Moonbase59 Okay, maybe not _that_ crazy.

I meant somewhere between what we have now in super-rudimentary form and fancier boolean searches, like... `artist = 'Taylor Swift' AND playlist IN ('shake-it-off-saturdays')`

I looked at QL's parser and...nah, I'm good

The idea would be to let these searches also be the foundation of playlists. i.e. you could say "This playlist is composed of the results of this search, refreshed periodically". Sort of a "smart" playlist.

**Reply by Moonbase59** [reply]
Problem might be that you can only restrict it to what tags are in the DB tables, otherwise tag reading/mapping would be much more complicated. (I can’t now even check for multiple genres, so possibilities would be exhausted for me real soon.)

It should still be possible to import a huge library like mine, without anything crashing. Might take time, of course.

Plus, ideally there should be means of saying "contains" and "is equal to", both with/without checking exact case (like QL’s appending a `c`).

From a programmer’s standpoint, I can see it makes you itch, hee hee. Go at it if you like… who knows what goodies might come out of it. Personally, I’d care about reading all kinds of tags first, _then_ use that as a base for more search possibilities. Must have a look at _getID3_ one of these days, see how much they’ve improved since I last used them.

**Reply by BusterNeece** [reply]
@Moonbase59 The search would just be what fields we have in our DB, so yeah, if you had custom tags on media that you wanted to incorporate into the search, it would need to be a custom field.
